src/System.Memory/src/System/MemoryExtensions.cs

```diff
@@ -10,13 +10,317 @@
+using System.Numerics;
+#endif
+
 namespace System
 {
     /// <summary>
     /// Extension methods for Span{T}, Memory{T}, and friends.
     /// </summary>
     public static partial class MemoryExtensions
     {
+        /// <summary>
+        /// Removes all leading and trailing white-space characters from the span.
+        /// </summary>
+        /// <param name="span">The span</param>
+        public static ReadOnlySpan<char> Trim(this ReadOnlySpan<char> span)
+        {
+            return span.Length == 0 ?
+                    default :
+                    span.TrimStart().TrimEnd();
+        }
+
+        /// <summary>
+        /// Removes all leading white-space characters from the span.
+        /// </summary>
+        /// <param name="span">The span</param>
+        public static ReadOnlySpan<char> TrimStart(this ReadOnlySpan<char> span)
+        {
+            int start = 0;
+            for (; start < span.Length; start++)
+            {
+                if (!char.IsWhiteSpace(span[start]))
+                    break;
+            }
+            return span.Slice(start);
+        }
+
+        /// <summary>
+        /// Removes all trailing white-space characters from the span.
+        /// </summary>
+        /// <param name="span">The span</param>
+        public static ReadOnlySpan<char> TrimEnd(this ReadOnlySpan<char> span)
+        {
+            int end = span.Length - 1;
+            for (; end >= 0; end--)
+            {
+                if (!char.IsWhiteSpace(span[end]))
+                    break;
+            }
+            return span.Slice(0, end + 1);
+        }
+
+        /// <summary>
+        /// Removes all leading and trailing occurrences of a specified character.
+        /// </summary>
+        /// <param name="span">The source span from which the character is removed.</param>
+        /// <param name="trimChar">The specified character to look for and remove.</param>
+        //[MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static ReadOnlySpan<char> Trim(this ReadOnlySpan<char> span, char trimChar)
+        {
+            return span.Length == 0 ?
+                    default :
+                    span.TrimStart(trimChar).TrimEnd(trimChar);
+        }
+
+        /// <summary>
+        /// Removes all leading occurrences of a specified character.
+        /// </summary>
+        /// <param name="span">The source span from which the character is removed.</param>
+        /// <param name="trimChar">The specified character to look for and remove.</param>
+        //[MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static ReadOnlySpan<char> TrimStart(this ReadOnlySpan<char> span, char trimChar)
+        {
+            if (span.Length == 0)
+                return default;
+
+            ref char first = ref MemoryMarshal.GetReference(span);
+            int start = 0;
+#if !netstandard11
+            if (Vector.IsHardwareAccelerated && span.Length >= 2 * Vector<ushort>.Count)
```

Comment:
Undo the vectorization?

Reply:
Should I undo the vectorization across the board, including `Trim(this ReadOnlySpan<char> span, ReadOnlySpan<char> trimChars)`?

What about parsing xml/json/code line by line and extracting comments from the sample example. It is possible to have deep nesting:
https://msdn.microsoft.com/en-us/library/system.string.trimstart(v=vs.110).aspx

Reply:
> Should I undo the vectorization across the board

I think so.

> It is possible to have deep nesting

Yes, it is possible that some people will be using this API to trim more than a few spaces. But it is not expected to be the common case we are optimizing for.